### PR TITLE
Add list parsing feature for gravity

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

This PR implements a list parsing feature into FTL that is intended to be used with gravity. Its main advantage is a stream-lined approach directly writing everything directly into the new database during analysis. There is no detour over huge allocated memory blocks or temporary files.

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

**Review by rdwebdesign — DISMISSED** [review]
Tested inside a container and it is working as expected.

**Review by rdwebdesign — APPROVED** [review]

File: src/gravity-tools.c [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+		total_read += read;
+		lineno++;
+
+		// Remove leading whitespace
+		// isspace() matches spaces, tabs, form feeds, line breaks,
+		// carriage returns, and vertical tabs
+		while(isspace(line[0]))
+			line++;
+
+		// Skip comments
```
Comment (yubiuser) [inline src/gravity-tools.c:166]: Here is some duplicate code L166-186. This is still part of `gravity.sh` 
`gravity_ParseFileIntoDomains()`

https://github.com/pi-hole/pi-hole/blob/73733308ba0f1c4e2cc5fb0c4aac6a9027349e48/gravity.sh#L639-L664

Does not hurt to keep it, just wanted to report.

Comment (DL6ER) [reply src/gravity-tools.c:166]: If it is duplicated, we should remove it from FTL for minimal speed gain.

Comment (DL6ER) [reply src/gravity-tools.c:166]: Actually, rethinking this, we can also remove `gravity_ParseFileIntoDomains()` from `gravity.sh` and use the code here that should be a lot faster.

Comment (yubiuser) [reply src/gravity-tools.c:166]: Then this would be something for https://github.com/pi-hole/pi-hole/pull/5275

**Comment by rdwebdesign** [comment]
@DL6ER 

I think I approved too soon.

I noticed the output here is saying `ignored 13 non-domain entries`, but only one entry is shown. 

```
  [i] Target: https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
  [✓] Status: Retrieval successful
  [✓] Parsed 183821 exact domains and 0 ABP-style domains (ignored 13 non-domain entries)
      Sample of non-domain entries:
        - "0.0.0.0"
```

Comparing to `master`, should be only 1 entry.
I think the other 12 lines are in the "false positive" list, but they are still counted.

**Comment by DL6ER** [comment]
@rdwebdesign This is expected and correct behavior. Each non-domain is counted but then only unique ones are shown. So, if there are 13 entries with `0.0.0.0` this is exactly what you are seeing here:

```
  [✓] Parsed 183821 exact domains and 0 ABP-style domains (ignored 13 non-domain entries)
      Sample of non-domain entries:
        - "0.0.0.0"
```

I'm always open to improving wording in case this isn't clear right now.

**Comment by yubiuser** [comment]
Maybe simply changing the message to

"Sample of non-domain entries (duplicates removed)"

**Pull request merged**